### PR TITLE
PIM-4985: redirection fixes on new navigation

### PR DESCRIPTION
**Pull request by sumbobyboys**
| Q                 | A
| ----------------- | ---
| Added Specs       | NA
| Added Behats      | NA
| Travis CI is ok   | No but it's normal, still WIP :)
| Changelog updated | N

### CODE REVIEW

File: features/association-type/edit_association_type.feature [hunk 1]
```diff
@@ -26,12 +26,11 @@ Feature: Edit an association type
       | title   | Are you sure you want to leave this page?                             |
       | content | You will lose changes to the association type if you leave this page. |
 
-  @skip
   Scenario: Successfully display a message when there are unsaved changes
     Given I am on the "PACK" association type page
     When I fill in the following information:
       | English (United States) | My pack |
-    Then I should see "There are unsaved changes."
+    Then I should see the text "There are unsaved changes."
```
Comment (jjanvier) [inline features/association-type/edit_association_type.feature]: nope, I disagree, we'll handle flash message later correctly

File: features/attribute-group/create_attribute_group.feature [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 
-  @skip
   Scenario: Successfully create an attribute group
     Given the "default" catalog configuration
     And I am logged in as "Julia"
     When I am on the attribute group creation page
     And I change the Code to "seo"
     And I save the attribute group
-    Then I should see "Attribute group successfully created"
+    Then I should see the text "Attribute group successfully created"
```
Comment (jjanvier) [inline features/attribute-group/create_attribute_group.feature]: same for all others

File: features/attribute-group/delete_attribute_group.feature [hunk 3]
```diff
@@ -14,16 +15,14 @@ Feature: Attribute group creation
       | red  | colors |
     And I am logged in as "Julia"
 
-  @javascript
   Scenario: Successfully delete an attribute group
     Given I am on the "sizes" attribute group page
     When I press the "Delete" button
     And I confirm the deletion
-    Then I should see flash message "Attribute group successfully removed"
+    Then I should see the text "Attribute group successfully removed"
```
Comment (jjanvier) [inline features/attribute-group/delete_attribute_group.feature]: same

File: features/attribute-group/delete_attribute_group.feature [hunk 4]
```diff
@@ -14,16 +15,14 @@ Feature: Attribute group creation
...
 
-  @javascript @skip
   Scenario: Fail to delete an attribute group that contains attributes
     Given I am on the "colors" attribute group page
     When I press the "Delete" button
     And I confirm the deletion
-    Then I should see flash message "Attribute group can't be removed as it contains attributes"
+    Then I should see the text "Attribute group can't be removed as it contains attributes"
```
Comment (jjanvier) [inline features/attribute-group/delete_attribute_group.feature]: same

File: features/association-type/edit_association_type.feature (hunk 1 above)
Comment (sumbobyboys) [inline features/association-type/edit_association_type.feature]: it's not a flash

File: features/attribute-group/create_attribute_group.feature (hunk 2 above)
Comment (sumbobyboys) [inline features/attribute-group/create_attribute_group.feature]: it was not a flash message check. Should I switch to a flash message check ?

File: features/Context/NavigationContext.php [hunk 5]
```diff
@@ -525,8 +525,12 @@ public function getPage($name)
      */
     public function iShouldBeRedirectedOnThePage($page)
     {
-        $page = isset($this->getPageMapping()[$page]) ? $this->getPageMapping()[$page] : $page;
-        $this->assertAddress($this->getPage($page)->getUrl());
+        $this->spin(function () use ($page) {
+            $page = isset($this->getPageMapping()[$page]) ? $this->getPageMapping()[$page] : $page;
```
Comment (phaseinducer) [inline features/Context/NavigationContext.php]: Could be outside of the spin imo

File: features/Context/NavigationContext.php [hunk 6]
```diff
@@ -665,8 +669,12 @@ public function iShouldBeOnTheCategoriesPage()
      */
     public function iShouldBeOnTheCategoryEditPage(Category $category)
     {
-        $expectedAddress = $this->getPage('Category edit')->getUrl(['id' => $category->getId()]);
-        $this->assertAddress($expectedAddress);
+        $this->spin(function () use ($category) {
+            $expectedAddress = $this->getPage('Category edit')->getUrl(['id' => $category->getId()]);
```
Comment (phaseinducer) [inline features/Context/NavigationContext.php]: same

File: features/Context/NavigationContext.php [hunk 7]
```diff
@@ -676,8 +684,12 @@ public function iShouldBeOnTheCategoryEditPage(Category $category)
      */
     public function iShouldBeOnTheCategoryNodeCreationPage(Category $category)
     {
-        $expectedAddress = $this->getPage('Category node creation')->getUrl(['id' => $category->getId()]);
-        $this->assertAddress($expectedAddress);
+        $this->spin(function () use ($category) {
+            $expectedAddress = $this->getPage('Category node creation')->getUrl(['id' => $category->getId()]);
```
Comment (phaseinducer) [inline features/Context/NavigationContext.php]: same

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/category.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    ], function (
+        $,
+        _,
+        mediator,
+        FormController,
+        router
+    ) {
+        return FormController.extend({
+            afterSubmit: function (xhr) {
+                this.$('#category-form').html(xhr.responseText);
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/category.js]: Will you generalize it later ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/registry.js [hunk 9]
```diff
@@ -1,22 +1,23 @@
...
 
-            require([controller], function (Controller) {
-                deferred.resolve(Controller);
-            });
-
-            return deferred.promise();
-        }
-    };
-});
+                return deferred.promise();
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/registry.js]: Why not returning from the require callback on line 16 ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-init.js [hunk 10]
```diff
@@ -135,9 +135,7 @@ define(
                             success: function () {
                                 loadingMask.hide().$el.remove();
                                 var targetUrl = $el.attr('data-redirect-url');
-                                // If already on the desired page, make sure it is refreshed
-                                Backbone.history.fragment = new Date().getTime();
-                                Backbone.history.navigate(targetUrl);
+                                router.redirect(targetUrl, {trigger: true});
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-init.js]: No spaces inside `{}` ?

**Comment by phaseinducer** [comment]
JSdoc on controllers would be nice. =)
Nice PR, much cleaner now !
GTM after jsdoc and jscs fix.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js [hunk 11]
```diff
@@ -1,31 +1,61 @@
...
+        TemplateController,
+        router
+    ) {
+        return TemplateController.extend({
+            events: {
+                'submit form': 'submitForm'
+            },
+            /**
+             * Handle form submission on the page
+             * @param {Event} event
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js]: new line please

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/base.js [hunk 12]
```diff
@@ -1,10 +1,16 @@
...
 
-define(function (require) {
-    var Backbone = require('backbone');
-
+define(['backbone'], function (Backbone) {
     return Backbone.View.extend({
+        /**
+         * Render the route given in parameter
+         *
+         * @param {String} route
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/base.js]: `string` please. `string` and `String` are different in js.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js [hunk 13]
```diff
@@ -1,31 +1,61 @@
...
+                    }.bind(this)
+                });
 
-            $form.ajaxSubmit({
-                complete: _.bind(function (xhr) {
+                return false;
+            },
+
+            /**
+             * Handle after submit success
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js]: Really not clear to me :/

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js [hunk 14]
```diff
@@ -1,31 +1,61 @@
...
+             *
+             * @param {Object} xhr
+             */
+            afterSubmit: function (xhr) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js]: Is it normal that there is no second parameter, but there's one on line 34 ?

**Comment by solivier** [comment]
GTM

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/registry.js [hunk 15]
```diff
@@ -1,22 +1,30 @@
...
+        var controllers       = module.config().controllers || {};
+        var defaultController = module.config().defaultController;
 
-    var controllers = module.config().controllers || {};
-    var defaultController = module.config().defaultController;
+        return {
+            /**
+             * Get the controller for the given name
+             *
+             * @param {String} name
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/registry.js]: same as above

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js (hunk 14 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js]: yup, just not used here

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js [hunk 16]
```diff
@@ -1,31 +1,61 @@
...
+        _,
+        mediator,
+        TemplateController,
+        router
+    ) {
+        return TemplateController.extend({
+            events: {
+                'submit form': 'submitForm'
+            },
+            /**
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/form.js]: new line please

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/template.js [hunk 17]
```diff
@@ -1,16 +1,23 @@
...
+    ['jquery', 'underscore', 'pim/controller/base'],
+    function ($, _, BaseController) {
+        return BaseController.extend({
+            /**
+             * {@inheritdoc}
+             */
+            renderRoute: function (route, path) {
+                return $.get(path).then(_.bind(this.renderTemplate, this)).promise();
+            },
+            /**
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/template.js]: new line

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/template.js [hunk 18]
```diff
@@ -1,16 +1,23 @@
...
+             * Add the given template to the current container
+             *
+             * @param {String} template
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/template.js]: same

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/variant-group.js [hunk 19]
```diff
@@ -17,6 +17,9 @@ define(
                 'submit form.form-horizontal': 'submitForm',
                 'submit form#pim_available_attributes': 'clearCache'
             },
+            /**
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/variant-group.js]: new line

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/variant-group.js [hunk 20]
```diff
@@ -39,6 +42,9 @@ define(
 
                 mediator.trigger('pim:reinit');
             },
+            /**
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/controller/variant-group.js]: new line

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/router.js [hunk 21]
```diff
@@ -1,117 +1,132 @@
...
-            this.loadingMask.render().$el.appendTo($('.hash-loading-mask'));
-            _.bindAll(this, 'showLoadingMask', 'hideLoadingMask');
+        var Router = Backbone.Router.extend({
+            DEFAULT_ROUTE: 'oro_default',
+            routes: {
+                '': 'dashboard',
+                '*path': 'defaultRoute'
+            },
+            loadingMask: null,
+            initialize: function () {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/router.js]: No jsdoc for this class ?

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**